### PR TITLE
core: add universal hardware support check function

### PR DESCRIPTION
**Pull request by tomoaki0705**
### This pullrequest changes
- use checkHalSimdSupport rather than calling checkHardwareSupport
- add SIMD check in spartialgradient.cpp

### CODE REVIEW

**Comment by alalek** [comment]
I agree with this unification.

Some notes:
1) It is "hal"-related function and I believe we should not garbage public API in cv namespace (actually the same note for other intrinsics - there is no problem to add "using namespace hal" into "namespace cv" of "core/hal/intrin.hpp" /cc @mshabunin @vpisarev ).
2) Currently this check is used for SIMD128. What is about this name: `cv::hal::hasSIMD128()`?
3) "inline" -> "static inline". Let's limit to internal linkage only.

**Comment by tomoaki0705** [comment]
Thank you for your review, @alalek .
Especially for (2), I really appreciate your suggestion since I'm really bad at naming.
Also, your comment remind me that I should also add ```cv::hal::hasSIMD128_64F()``` since only ARMv8 accepts double as an element on ARM platform.

**Comment by tomoaki0705** [comment]
I followed 2) and 3).
For 1), I'd like to wait the conclusion of discussion.

Best, regards.

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp [hunk 1]
```diff
@@ -1206,6 +1207,16 @@ inline v_float16x4 v_cvt_f16(const v_float32x4& a)
 }
 #endif
 
+//! @name Check SIMD support
+//! @{
+//! @brief Check CPU capability of SIMD operation
+static inline bool hasSIMD128()
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/intrin_neon.hpp]: shall we also use predefined macros like `__ARM_NEON__` e.g. like:

<pre>
#ifdef __ARM_NEON__
    return true;
#else
   return checkHardwareSupport(CV_CPU_NEON);
#endif
</pre>

File: modules/core/include/opencv2/core/hal/intrin_sse.hpp [hunk 2]
```diff
@@ -1686,6 +1687,16 @@ inline v_float16x4 v_cvt_f16(const v_float32x4& a)
 }
 #endif
 
+//! @name Check SIMD support
+//! @{
+//! @brief Check CPU capability of SIMD operation
+static inline bool hasSIMD128()
+{
+    return checkHardwareSupport(CV_CPU_SSE2);
```
Comment (vpisarev) [inline modules/core/include/opencv2/core/hal/intrin_sse.hpp]: same comment. Shall we use predefined macro?

File: modules/core/include/opencv2/core/hal/intrin_neon.hpp (hunk 1 above)
Comment (tomoaki0705) [inline modules/core/include/opencv2/core/hal/intrin_neon.hpp]: I don't get the point.
NEON is not a part of instruction. It's an extention, so checking the hardware support has a meaning.
Why do we need to return ```true``` using macro?

Comment (alalek) [inline modules/core/include/opencv2/core/hal/intrin_neon.hpp]: @vpisarev I believe we should allow to disabling optimizations during runtime (so we should call `checkHardwareSupport()`).

**Comment by vpisarev** [comment]
ok, let's put it in.  it's useful change and we can refine it later :+1:

**Pull request merged**